crates/utils/signposter/src/lib.rs

```diff
@@ -0,0 +1,204 @@
+
+#[cfg(feature = "global")]
+pub use global::*;
+
+pub enum LogCategory {
+    /// Signposts provide high-level events that can be used to orient a
+    /// developer looking at performance data. These will be displayed by
+    /// default by performance tools like Instruments.app.
+    PointsOfInterest,
+    /// Signposts should be disabled by default, reducing the runtime overhead.
+    /// [`Log::signpost_enabled`] calls will only return 'true' when a
+    /// performance tool like Instruments.app is recording.
+    DynamicTracing,
+    /// Signposts should capture user backtraces. This behavior is more
+    /// expensive, so [`Log::signpost_enabled`] will only return 'true' when
+    /// a performance tool like Instruments.app is recording.
+    DynamicStackTracing,
+}
+
+impl LogCategory {
+    fn as_c_str(&self) -> &CStr {
+        match self {
+            LogCategory::PointsOfInterest => {
+                CStr::from_bytes_with_nul(b"PointsOfInterest\0").unwrap()
+            }
+            LogCategory::DynamicTracing => CStr::from_bytes_with_nul(b"DynamicTracing\0").unwrap(),
+            LogCategory::DynamicStackTracing => {
+                CStr::from_bytes_with_nul(b"DynamicStackTracing\0").unwrap()
+            }
+        }
+    }
+}
+
+/// A log is a collection of signposts.
+#[derive(Debug, Clone, Copy)]
+pub struct Log {
+    os_log: *mut os_log_s,
+}
+
+// Safety: unclear
+unsafe impl Sync for Log {}
+// Safety: unclear
+unsafe impl Send for Log {}
+
+impl Default for Log {
+    fn default() -> Self {
+        Log {
+            os_log: unsafe { &_os_log_default as *const _ as *mut _ },
+        }
+    }
+}
+
+impl Log {
+    /// Create a new log with the given subsystem and category.
+    pub fn new_with_subsystem_and_category(subsystem: &str, category: LogCategory) -> Self {
+        let subsystem = CString::new(subsystem).unwrap();
+        let category = category.as_c_str();
+        let os_log = unsafe { os_log_create(subsystem.as_ptr(), category.as_ptr()) };
+        Log { os_log }
+    }
+
+    /// Create a new [`Signpost`] for this log.
+    pub fn signpost(&self) -> Signpost {
+        let id = unsafe { os_signpost_id_generate(self.os_log) };
+        Signpost {
+            id,
+            log: self.os_log,
+        }
+    }
+}
+
+/// A signpost is a marker that can be used to annotate a log with
+/// information about the execution of a program.
+#[derive(Debug, Clone, Copy)]
+pub struct Signpost {
+    id: os_signpost_id_t,
+    log: *mut os_log_s,
+}
+
+// Safety: unclear
+unsafe impl Sync for Signpost {}
+// Safety: unclear
+unsafe impl Send for Signpost {}
+
+// NOTE(alexkirsz) Not sure that's correct, but it's what the Dart SDK does and
+// it seems to work.
+// See https://github.com/dart-lang/sdk/blob/3e2d3bc77fa8bb5139b869e9b3a5357b5487df18/runtime/vm/timeline_macos.cc#L34-L35
+static FORMAT_BUFFER: [u8; 64] = [0; 64];
+static FORMAT_BUFFER_LEN: u32 = (FORMAT_BUFFER.len() * std::mem::size_of::<u8>()) as u32;
```

Comment:
I think this can be a constant, but I'm not sure if it matters

Reply:
I can always hardcode it to 64 and use it in `FORMAT_BUFFER`'s initialization.